### ISSUE TITLE
Feature request: Allow overriding untemplated values in templates

### ISSUE THREAD

**Issue by omkensey** [helm/helm#2399]
It would be nice if I could override values in templates even if they're not properly templated, e.g. given this:

templates/myservice.yaml:
```
apiVersion: v1
kind: Service
metadata:
  labels:
    app: myapp
    component: frontend
  name: myapp-frontend-svc
spec:
  ports:
    - name: myapp-frontend
      port: 80
      targetPort: 8080
      protocol: TCP
  selector:
    app: myapp
    component: frontend
```

I could do something like this:

`$ helm install stable/myapp --dynamic-value Values.Local.servicePort=myservice.yaml:spec.ports.myapp-frontend.port --set Values.Local.servicePort=81`

(N.B. Kubernetes doesn't require names if a service only has one port, in which case :spec.ports.port could be valid.)

An alternate method could be to simply assign values to the parse tree endpoint directly, e.g. `--set-dynamic myservice.yaml:spec.ports.myapp-frontend.port=81`

This would probably require annotating the deployed object in some way so that future helm sessions can deal intelligently with dynamically-set values that don't correspond to any known version of a chart.

**Comment by thomastaylor312** [comment]
@omkensey Thank you for the feature request. Ideally the proper way to do this is to templatize things. Could you explain a current use case to clarify this some more?

@technosophos This feels like something that may have been discussed before (whether on or off record). Do you recall anything about this?

**Comment by thomastaylor312** [comment]
I am labeling this as question/support until we flesh it out some more

**Comment by omkensey** [comment]
I'm thinking of a case where a chart exists, but doesn't template something that a user needs templated.  Ideally that results in a PR from the user, but a) not everybody has the ability or permission to do that, b) until that PR is acted on and hopefully merged, they still have to get work done -- which means either using the chart as-is and writing additional customization outside it, or altering it and maintaining it locally.  Dynamic override would be a stopgap/intermediate where the chart remains useful as-is (also allowing the user to benefit from any additional templating done on it for other things) while not requiring so much additional workload to be carried on their plate.

**Comment by scottrigby** [comment]
This is something I brought up on yesterday's charts dev call, and @viglesiasce pointed me to this proposal, which I have not yet read but so far sounds very similar: https://docs.google.com/document/d/1cLPGweVEYrVqQvBLJg6sxV-TrE5Rm2MNOBA_cxZP2WU/edit#heading=h.dr88tktf0e99 I know Today we have Go Template, and this idea would be a long way off if it happens - is there a more canonical issue for discussing this?

**Comment by thomastaylor312** [comment]
@scottrigby I don't think there is a more canonical issue yet. If you have the desire/time to write a full proposal, we could close this issue and use the new one as the main discussion

**Comment by Nitrogenycs** [comment]
I just ran into this problem with the swift chart at https://github.com/kubernetes/charts/blob/master/stable/swift/templates/service.yaml . I'd like to turn it into a NodePort service, but this part is not templated.

So I see 4 possible ways to solve this problem:

1. In this special case I could probably make a 2nd service with NodePort enabled that selects the swift app.
2. I fetch the swift dependency and monkey-patch things myself.
3. I can override untemplated parts as outlined by @omkensey.
4. I fix the swift template.

The solution that feels most clean to me is 4. However if I wanted to override something more esoteric than the service type or if the original chart author did not want to take my patch then I'd be back at something like 2 or 3.

In conclusion, the solution described by @omkensey would solve some real world use cases if fixing the template is not an option. Ofc you could achieve the same by using dependency fetch and then applying a good old patch :-)

**Comment by szwed** [comment]
YAML also provides a handy feature called **anchors**, which let you easily duplicate
content across your document. Both of these keys will have the same value:

```
default: &default_title This Post Has No Title
title: *default_title
```

maps to:

```
18:40 $ helm install --dry-run --debug someapp/ |grep -i this
default: This Post Has No Title
title: This Post Has No Title
```

Doesn't allow for concatenation or more advanced combinations, but in some cases even this limited functionality may help a lot.

**Comment by nick4fake** [comment]
Another example:
https://github.com/kubernetes/charts/pull/3894

There is no way in Jenkins HELM package to set resource limits ATM

**Comment by nick4fake** [comment]
Forking every helm chart to add tolerations is a nightmare. Are there any workarounds?

**Comment by viglesiasce** [comment]
@nick4fake You may want to look at using [Kustomize](https://github.com/kubernetes-sigs/kustomize) to do the "last-mile" modifications for things that dont exist in the charts but should be applied broadly in your setup. Kustomize allows you to add overlays on top of manifests. You could do something like:

```shell
helm template -f values.yaml stable/jenkins > forked-chart/base.yaml
# Add customization in forked-chart/tolerations.yaml
kustomize build forked-chart | kubectl apply -f -
```

**Comment by bacongobbler** [comment]
I would also refer to some of the existing proposals in Helm 3, in particular around extending existing charts and the event model changes:

https://github.com/kubernetes-helm/community/blob/master/helm-v3/001-charts.md
https://github.com/kubernetes-helm/community/blob/master/helm-v3/002-events.md

**Comment by johnmarcou** [comment]
+1 on this FR, it would be awesome!

It's always a nightmare to hack chart which doesn't support yet this or that feature we need.

**Comment by AndriiNeverov** [comment]
+1

**Comment by TheFrozenFire** [comment]
:+1:

**Comment by bacongobbler** [comment]
related: #5276

**Comment by michelleN** [comment]
This is possible now using post render feature + kustomize as of helm 3.1
@thomastaylor312's example is located [here](https://github.com/thomastaylor312/advanced-helm-demos/tree/master/post-render)
Docs on the post render feature is located [here](https://helm.sh/docs/topics/advanced/).

Please re-open the issue if something was missed.

**Comment by eledhwen** [comment]
I think this is very unfortunate that this requires separate tools (either kustomize or separate scripts) for something as common as this. I understand the appeal behind the ability to use different operators for altering the output, but a default mechanism should certainly be provided for managing such Chart alterations.

I suppose forking charts will have to do for now.
